File: Android/tuilivekit/livekit/src/main/java/com/trtc/uikit/livekit/features/livelist/view/singlecolumn/LiveStatusView.java
package com.trtc.uikit.livekit.features.livelist.view.singlecolumn;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.LinearInterpolator;

import androidx.annotation.NonNull;

public class LiveStatusView extends View {
    private static final int BAR_COUNT     = 3;
    private static final int BAR_WIDTH     = 2;
    private static final int BAR_HEIGHT    = 7;
    private static final int BAR_SPACING   = 3;
    private static final int ANIM_DURATION = 1000;
    private static final int COLOR         = Color.parseColor("#FFFFFF");

    private       Paint         mBarPaint;
    private final float[]       mBarHeights = new float[3];
    private       ValueAnimator mAnimator;

    public LiveStatusView(Context context) {
        super(context);
        init();
    }

    public LiveStatusView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        mBarPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mBarPaint.setColor(COLOR);
        mBarPaint.setStyle(Paint.Style.FILL);

        for (int i = 0; i < BAR_COUNT; i++) {
            mBarHeights[i] = dp2px(BAR_HEIGHT);
        }
    }

    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);

        int totalWidth = dp2px(BAR_WIDTH * BAR_COUNT + BAR_SPACING * (BAR_COUNT - 1));
        int startX = (getWidth() - totalWidth) / 2;
        int centerY = getHeight() / 2;

        for (int i = 0; i < BAR_COUNT; i++) {
            float left = startX + i * (dp2px(BAR_WIDTH + BAR_SPACING));
            float top = centerY - mBarHeights[i] / 2;
            float right = left + dp2px(BAR_WIDTH);
            float bottom = centerY + mBarHeights[i] / 2;

            canvas.drawRoundRect(left, top, right, bottom, dp2px(2), dp2px(2), mBarPaint);
        }
    }

    private void startAnimation() {
        if (mAnimator != null && mAnimator.isRunning()) {
            mAnimator.cancel();
        }

        mAnimator = ValueAnimator.ofFloat(0, 1);
        mAnimator.setDuration(ANIM_DURATION);
        mAnimator.setRepeatCount(ValueAnimator.INFINITE);
        mAnimator.setRepeatMode(ValueAnimator.REVERSE);
        mAnimator.setInterpolator(new LinearInterpolator());

        mAnimator.addUpdateListener(animation -> {
            float progress = (float) animation.getAnimatedValue();
            for (int i = 0; i < BAR_COUNT; i++) {
                float phase = i * 0.6f;
                float adjustedProgress = (progress + phase) % 1.0f;
                mBarHeights[i] = dp2px(BAR_HEIGHT + (float) (Math.sin(adjustedProgress * Math.PI) * 15));
            }
            invalidate();
        });

        mAnimator.start();
    }

    public void stopAnimation() {
        if (mAnimator != null) {
            mAnimator.cancel();
        }
    }

    private int dp2px(float dp) {
        return (int) (dp * getResources().getDisplayMetrics().density + 0.5f);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        startAnimation();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        stopAnimation();
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/features/livelist/view/singlecolumn/LiveStatusView.kt
package com.trtc.uikit.livekit.features.livelist.view.singlecolumn

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import android.view.animation.LinearInterpolator
import androidx.core.graphics.toColorInt
import kotlin.math.sin

class LiveStatusView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    companion object {
        private const val BAR_COUNT = 3
        private const val BAR_WIDTH = 2
        private const val BAR_HEIGHT = 7
        private const val BAR_SPACING = 3
        private const val ANIM_DURATION = 1000
        private val COLOR = "#FFFFFF".toColorInt()
    }

    private val barPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = COLOR
        style = Paint.Style.FILL
    }
    private val barHeights = FloatArray(BAR_COUNT) { dp2px(BAR_HEIGHT.toFloat()) }
    private var animator: ValueAnimator? = null

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val totalWidth = dp2px((BAR_WIDTH * BAR_COUNT + BAR_SPACING * (BAR_COUNT - 1)).toFloat())
        val startX = (width - totalWidth) / 2
        val centerY = height / 2

        repeat(BAR_COUNT) { i ->
            val left = startX + i * dp2px((BAR_WIDTH + BAR_SPACING).toFloat())
            val top = centerY - barHeights[i] / 2
            val right = left + dp2px(BAR_WIDTH.toFloat())
            val bottom = centerY + barHeights[i] / 2

            canvas.drawRoundRect(left, top, right, bottom, dp2px(2f), dp2px(2f), barPaint)
        }
    }

    private fun startAnimation() {
        animator?.takeIf { it.isRunning }?.cancel()

        animator = ValueAnimator.ofFloat(0f, 1f).apply {
            duration = ANIM_DURATION.toLong()
            repeatCount = ValueAnimator.INFINITE
            repeatMode = ValueAnimator.REVERSE
            interpolator = LinearInterpolator()

            addUpdateListener { animation ->
                val progress = animation.animatedValue as Float
                repeat(BAR_COUNT) { i ->
                    val phase = i * 0.6f
                    val adjustedProgress = (progress + phase) % 1.0f
                    barHeights[i] = dp2px(BAR_HEIGHT + (sin(adjustedProgress * Math.PI) * 15).toFloat())
                }
                invalidate()
            }
            start()
        }
    }

    fun stopAnimation() {
        animator?.cancel()
    }

    private fun dp2px(dp: Float): Float {
        return (dp * resources.displayMetrics.density + 0.5f)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        startAnimation()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        stopAnimation()
    }
}